### ISSUE TITLE
allow "uv run" to work without a project

### ISSUE THREAD

**Issue by mmerickel** [astral-sh/uv#8666]
It'd be nice if this flow worked simply to avoid needing to activate a virtualenv - regardless of whether I'm using a full uv project or not.

```
$ uv venv
$ uv pip install -e .
$ uv run python
```

However as of 0.4.28 I get the following error: 

```
$ uv run python
error: No `project` table found in: `/Users/michael/work/oss/webob/pyproject.toml`
```

Obviously in this case I'm not asking uv to do anything except use the correct path as if I was activating the virtualenv - equivalent to `.venv/bin/activate; python`.

**Comment by mmerickel** [comment]
Yeah thanks for pointing that out and I confirmed it works. Having to add an extra arg there on every command is a deal breaker right - `uv run` is the primary command I’m using all the time during dev. I’d instead activate the env or use `.venv/bin/python …` (my current flow). I guess this issue becomes a feature request to drop the flag and make it the default fallback behavior when no project can be found.

**Comment by mmerickel** [comment]
Things like build-system, black config, isort config, and tool.ruff config.

**Comment by zanieb** [comment]
What build system are you using without a `[project]` table 👀

**Comment by mmerickel** [comment]
setuptools doesn’t require it - it can pull all the data from setup.py and setup.cfg like it has for decades. :-) I guess I should switch it up to use project but it’s not required by the build backend PEP.

**Comment by samypr100** [comment]
> Having to add an extra arg there on every command is a deal breaker right - uv run is the primary command I’m using all the time during dev

Maybe as an alternative workaround, setup an shell alias for `uv run` to be `uv run --no-project`?

**Comment by mmerickel** [comment]
Yeah I can work around it, but this ticket is here incase y'all want a cleaner story for the larger community for "older" projects.

In pyramid we have for years recommended for people to do the following as a simple starter sequence:

```
python3 -m venv env
env/bin/pip install -e .
env/bin/pshell site.ini
```

And I thought it'd be cool to at least have a "drop in replacement" story for uv where we'd recommend the following:

```
uv venv
uv pip install -e .
uv run pshell site.ini
```

As of today we need to add `--no-project` to every `uv run` command and that's no fun. :-) Personally I think the error that is raised about "no project table" is not worth it compared to the logical fallback to just run the command in the virtualenv even if there's no project.

Also mainly for my sanity and to clarify for anyone else I went back to just check and this issue occurs whether or not you have a pyproject.toml. It's just complaining about the missing `[project]` section therein.

**Comment by zanieb** [comment]
Fwiw I think this should probably work but doesn't

```
❯ cat pyproject.toml
[tool.uv]
managed = false

❯ uv run python
error: No `project` table found in: `/Users/zb/workspace/uv/example/pyproject.toml`
```

**Comment by mattp-** [comment]
👍 on this, I use uv primarily as a manager to venv's per project, but im not necessarily making changes to the projects im dropping into that would allow me to modify pyproject.toml. Being able to set an env var that silences no-project, but still works when a project dir i go to actually DOES have project settings, would be ideal.

**Comment by matterhorn103** [comment]
If a directory contains nothing but a script, with no `pyproject.toml` or `.venv` in sight, `uv run script.py` works.

If there's a `.venv` but no `pyproject.toml`, it also works.

Is that the intended behaviour? If so it seems strange to require a `[project]` table when `pyproject.toml` is present, but not otherwise.

**Comment by zanieb** [comment]
The presence of a `pyproject.toml` is what we use to determine if you're working... in a project :) I think it does make some sense, though I can understand why you would want it to work differently. I'll be exploring some approaches to resolve this.

**Comment by trim21** [comment]
> What build system are you using without a `[project]` table 👀

I a example for this, a python backend web server doesn't need any build-system. When deploying I just need to copy everything to docker and install deps from requirements.txt, no `[project]` involved.

Maybe we can add a options or env to make `--no-project` enabled by default.

For example, `UV_RUN_NO_PROJECT=1 uv run ... ` ( even uv doesn't load dotenv for itself direnv can still work in this case ) or `tool.uv.no-project = true` in config file

**Comment by trim21** [comment]
I just find that we have a option `tool.uv.package`, maybe we can use this option when `tool.uv.package = false`?

https://github.com/astral-sh/uv/blob/8a27d4d340f7dbdaeb144d4b8f80f3b27cc29ec0/docs/concepts/projects/config.md?plain=1#L147-L148

**Comment by Avasam** [comment]
For real world examples, I've been hitting this in typeshed. Not sure if I wanna try adding a dummy project config.

**Comment by stefanv** [comment]
In trying to use uv to prepare environments for a web application (no compilation or package involved), I noticed that `uv run` gives:

```
error: No `project` table found in: `<somepath>/pyproject.toml`
```

OK, so we add the project section to `pyproject.toml`. Now we get:

```
$ uv run make
  × Failed to build `skyportal @ file:///home/stefan/src/skyportal`
  ├─▶ The build backend returned an error
  ╰─▶ Call to `setuptools.build_meta:__legacy__.build_wheel` failed (exit status: 1)

      [stderr]
      error: Multiple top-level packages discovered in a flat-layout: ...
```

OK, so it seems I *have* to run `uv run --no-project`. Which, presumably should have been the case originally when the `project` section was missing. Having a way to configure this behavior, without a manual flag on each invocation, would be very helpful.

**Comment by helderco** [comment]
Here's an example I ran into today: https://github.com/swe-bench/SWE-bench

It has a `pyproject.toml` but with only the following:
```toml
[build-system]
requires = ['setuptools>=42']
build-backend = 'setuptools.build_meta'
```
Notice the above doesn't include a `[project]` section, but you can install it with `pip install -e .`. There's a [`setup.py`](https://github.com/swe-bench/SWE-bench/blob/a0536ee6f9fd5ff88acf17a36a384bf3da3d93d6/setup.py) where the dependencies are declared. 

My team member is having issues with his environment, so I wish I could tell him to just slap `uv run` in front:
```
❯ uv run -m swebench.harness.run_evaluation \
    --predictions_path gold \
    --max_workers 1 \
    --instance_ids sympy__sympy-20590 \
    --run_id validate-gold
error: No `project` table found in: `/.../SWE-bench/pyproject.toml`
```
Don't want to instruct on changing that open source project's `pyproject.toml` to accomodate, and wish I wouldn't have to recommend going through the `uv pip` route.

**Comment by charliermarsh** [comment]
👍 We do currently plan to support this -- hopefully soon.

**Comment by rbavery** [comment]
another example https://github.com/567-labs/systematically-improving-rag/tree/main/cohort_2

**Comment by T-256** [comment]
Does #13742 solve this issue?

**Comment by mmerickel** [comment]
I just tested with the current uv main commit and the sequence of commands in the ticket result in the same error so no it doesn't solve the issue.
